### PR TITLE
8300109: RISC-V: Improve code generation for MinI/MaxI nodes

### PR DESCRIPTION
**Pull request by RealFYang**
As described by the issue, code generation for MinI/MaxI nodes on RISC-V could be improved when one of the source register is the same as the destination register. Also the code could be further simplified when that source register is constant 0 making use of the dedicated zero register of the architecture. This adds new match rules for those cases. 

Testing:
- [x] Run the new jtreg test added by: https://github.com/openjdk/jdk/pull/11570
- [x] Tier1-3 tested with release build on HiFive Unmatched board.
- [x] Run non-trivial benchmark workloads (dacapo, specjvm, renaissance, etc) with fastdebug build on HiFive Unmatched board.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300109](https://bugs.openjdk.org/browse/JDK-8300109): RISC-V: Improve code generation for MinI/MaxI nodes

### Reviewers
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author) ⚠️ Review applies to [a17a391c](https://git.openjdk.org/jdk/pull/11988/files/a17a391cd9bc0ba4e368723c238bd304158caf55)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) ⚠️ Review applies to [a17a391c](https://git.openjdk.org/jdk/pull/11988/files/a17a391cd9bc0ba4e368723c238bd304158caf55)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [a17a391c](https://git.openjdk.org/jdk/pull/11988/files/a17a391cd9bc0ba4e368723c238bd304158caf55)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11988/head:pull/11988` \
`$ git checkout pull/11988`

Update a local copy of the PR: \
`$ git checkout pull/11988` \
`$ git pull https://git.openjdk.org/jdk pull/11988/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11988`

View PR using the GUI difftool: \
`$ git pr show -t 11988`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11988.diff">https://git.openjdk.org/jdk/pull/11988.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/riscv.ad [hunk 1]
```diff
@@ -8657,6 +8657,94 @@ instruct cmpLTMask_reg_zero(iRegINoSp dst, iRegIorL2I op, immI0 zero)
...
+  %}
+
+  ins_pipe(pipe_class_default);
+%}
+
+// special case for comparing with zero
+// n.b. this is selected in preference to the rule above because it
+// avoids loading constant 0 into a source register
+
+instruct minI_reg_zero(iRegINoSp dst, immI0 zero)
```
Comment (dchuyko) [inline src/hotspot/cpu/riscv/riscv.ad:8706]: As mentioned in https://github.com/openjdk/jdk/pull/11570, unfortunately the order of immediate in MinI/MaxI is not quaranteed, at least it is so in Ideal().

Comment (RealFYang) [reply src/hotspot/cpu/riscv/riscv.ad:8706]: Yes, I think you are right. Added another two match rules for that case.

**Review by feilongjiang — APPROVED** [review]

**Review by luhenry — APPROVED** [review]
I'm curious to see some JMH results on this change.

**Review by shipilev — APPROVED** [review]
Looks okay. I was wondering if it was possible to have multiple `match(...)`-es per `instruct`, so we could coalesce the symmetric "compare with zero" cases. But it is fine as it is.

**Comment by RealFYang** [comment]
> I'm curious to see some JMH results on this change.

Well, not very much (%3 at most) when I run the max & min tests in JMH test [1] on my unmatched board:

```
Before:
Benchmark          (seed)   Mode  Cnt       Score      Error   Units
MathBench.maxInt        0  thrpt    8   97675.352 ± 4081.422  ops/ms
MathBench.maxLong       0  thrpt    8  105212.075 ± 5305.248  ops/ms
MathBench.minInt        0  thrpt    8  104321.539 ± 4342.072  ops/ms
MathBench.minLong       0  thrpt    8   97597.538 ± 3798.621  ops/ms

After:
Benchmark          (seed)   Mode  Cnt       Score      Error   Units
MathBench.maxInt        0  thrpt    8   97890.688 ± 6732.534  ops/ms
MathBench.maxLong       0  thrpt    8  107836.865 ± 3148.886  ops/ms
MathBench.minInt        0  thrpt    8  107521.081 ± 3992.511  ops/ms
MathBench.minLong       0  thrpt    8   97940.617 ± 3876.168  ops/ms
```
I think it will depend on the micro-architecture implementation.

[1] https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/java/lang/MathBench.java#L301

**Comment by RealFYang** [comment]
Thanks all for the review. I will proceed integrating this if there is no other comments.

**Pull request closed without merging**